---
output: pdf_document
---
# 16S Project Report
```{r setup, echo=FALSE}
#Creates figure image files in Knit folder
library(knitr)
opts_chunk$set(
  tidy=FALSE,
  cache=T,
  echo=F,
  dpi=100,
  fig.width=6,
  fig.path="Knit/figure/",
  cache.path="Knit/cache/",
  dev=c("png", "pdf"))
```

```{r load scripts, echo=FALSE, message=FALSE}
#Load scripts for 16S analysis
library(qiimer)
library(pander)
library(ape)
library(vegan)
library(ggplot2)
library(plyr)
#library(dplyr)
library(reshape2)
#library(kylemisc)
```

```{r define_functions}
make_pcoa_plot <- function(uu, s, shape_by, color_by) {
  uu <- dist_subset(uu, s$SampleIDorig)
  uu_pcoa <- pcoa(uu)
  uu_df <- cbind(s, uu_pcoa$vectors[s$SampleIDorig, 1:5])
  uu_pct <- round(uu_pcoa$values$Relative_eig * 100)

  g_uu = ggplot(uu_df, aes(x=Axis.1, y=Axis.2)) +
    coord_equal() +
    theme_bw() +
    xlab(paste0("PCoA axis 1 (", uu_pct[1], "%)")) +
    ylab(paste0("PCoA axis 2 (", uu_pct[2], "%)")) +
    scale_shape_discrete(name=shape_by) + 
    scale_colour_discrete(name=color_by) +
    ggtitle('PCoA plot based on unweighted unifrac distances')

  if (shape_by =='' & color_by!='') {
    g_uu <- g_uu + geom_point(aes(colour=factor(get(color_by))))
  } else if (shape_by!='' & color_by!='') {
    g_uu <- g_uu + geom_point(aes(colour=factor(get(color_by)), shape=factor(get(shape_by))))
  } else {
    g_uu <- g_uu + geom_point()
  }
  return(g_uu)
}
```


```{r define_constants code, echo=FALSE}
# User defined variables
work_dir <- "/media/THING1/tanesc/friedman/SparseFur"
mapping_file_name <- "SparseFur_sampleSheet.tsv"

# File paths
mapping_file_fp <- file.path(work_dir, mapping_file_name)
otu_table_fp <- file.path(work_dir, "otu", "otu_table.txt")
split_library_log_fp <- file.path(work_dir, "library", "split_library_log.txt")
uu_fp <- file.path(work_dir, "beta_diversity", "unweighted_unifrac_dm.txt")
wu_fp <- file.path(work_dir, "beta_diversity", "weighted_normalized_unifrac_dm.txt")

# Run parameters
min_reads <- 1000
```

```{r DeNOVO OTU code, echo=FALSE}
# Assign sample mapping file
s <- read_qiime_mapping_file(mapping_file_fp)

# Assign OTU table
o <- read_qiime_otu_table(otu_table_fp)

# Metadata in the form of truncated green genes assignments
md <- sub("(; [kpcofgs]__)+$", "", o$metadata, perl=T)

# Assignments data-frame
adf <- split_assignments(md)
a <- simplify_assignments(adf)

# Check if the sample names start with a number. If they do, add an "X" to the front
toFix = grepl("^[0-9]", s$SampleID)
s$SampleIDorig = s$SampleID
s$SampleID[toFix] = paste0('X', s$SampleID[toFix])

# Check for invalid sample IDs
problem_ids <- setdiff( colnames(o$counts), s$SampleID)
if (length(problem_ids ) > 0) stop (simpleError(paste("id mismatch found", problem_ids, collapse=" ")))

# check for the column names to assign color_by and shape_by for pcoa plots
color_by = ''
shape_by = ''
potential_headers <- c('study_group', 'study_day', 'current_antibiotics', 'cage_number', 'mouse_strain')
header_idx = which(is.element(potential_headers, colnames(s)))

if(length(header_idx)>0){
  color_by <- potential_headers[header_idx[1]]
}
if(length(header_idx)>1){
  shape_by <- potential_headers[header_idx[2]]
}
```

```{r echo=FALSE}
# Assignment counts per SampleID
cts <- o$counts[,s$SampleID]
```

## --------------------------------------------------------

## Histogram of high quality paired reads per sample
The vertical line shows the minimum number of reads for analysis

```{r histogram, fig.width=7, fig.height=5, echo=FALSE}
read_counts <- read.csv(split_library_log_fp, sep="\t", skip=15, strip.white = FALSE, blank.lines.skip = TRUE,  header=FALSE, col.names=c("SampleID", "Read Counts"), stringsAsFactors = FALSE)

read_counts <- read_counts[1:(nrow(read_counts)-2),]
read_counts <- read_counts[order(read_counts$SampleID), ]
rownames(read_counts) <- NULL

toFix = grepl("^[0-9]", read_counts$SampleID)
read_counts$SampleID[toFix] = paste0('X', read_counts$SampleID[toFix])

s = merge(s, read_counts, by="SampleID", all.x = TRUE)

ggplot(s, aes(x=Read.Counts)) +
    geom_histogram(binwidth=10000) +
    geom_vline(xintercept = min_reads) +
    theme_classic() +
    theme_bw() + 
    xlab("Number of reads in sample") +
    ylab("Number of samples")

```

## Whole samples that are above the 1000 read count threshold
```{r}
## Whole samples discarded that are below the read count threshold of 1000
s$Keep <- s$Read.Counts > min_reads
s$KeepLabel <- factor(ifelse(s$Keep, "Keep", "Discard"))
keep_table <- table(s[,color_by], s$KeepLabel)
pander(keep_table,caption = "samples lost during prep and sequencing")
```

## Taxonomic Heatmap
The most specific taxonomic assignment available for each read is shown. Taxa are included in the plot if more than 1000 observations are made across the entire dataset. "Unassigned" reads did not match any 16S sequence in the reference database at a similarity threshold of 90%
Taxa are grouped by lowest common ranks for each OTU. Ranks are included in the plot if the combined OTUs reach 1000 observations together.
```{r overall_heatmap_annotations, echo=FALSE}
# s <- arrange(s, Project, Study_Group, Study_Day)
# s$Study_Group <- as.factor(as.character(s$Study_Group))
# s$Study_Day <- as.factor(s$Study_Day)
# annotations <- s[,c("Project", "Study_Group","SubjectID", "SampleType", "Study_Day")]
# rownames(annotations) <- s$SampleID
```

```{r Heatmap_by_study, fig.width=24, fig.height=15, echo=FALSE}
otu_heatmap(
  cts,
  a,
  #annotation=annotations,
  threshold=1000,
  color = saturated_rainbow(max(colSums(cts)), saturation_limit = 0.50),
  cluster_cols = FALSE, cluster_rows = FALSE,
  fontsize_col = 10, fontsize_row = 10,
  cellheight=10, cellwidth=10,
  legend = TRUE)

```


## Beta diversity measures
```{r Ordination_plots_uu, fig.width=7, fig.height=5, echo=FALSE}
uu <- read_qiime_distmat(uu_fp)
g_uu <- make_pcoa_plot(uu, s, shape_by, color_by)
plot(g_uu)
```

```{r Cluster_by_study_uu, fig.width=7, fig.height=7, echo=FALSE}
hc = hclust(uu)
plot(hc, main="Complete linkage clustergram based on\nunweighted unifrac distances", xlab="", ylab="distance", sub = "")
```

```{r Ordination_plots_wu, fig.width=7, fig.height=5, echo=FALSE}
wu <- read_qiime_distmat(wu_fp)
g_wu <- make_pcoa_plot(wu, s, shape_by, color_by)
plot(g_wu)
```

```{r Cluster_by_study_wu, fig.width=7, fig.height=7, echo=FALSE}
hc = hclust(wu)
plot(hc, main="Complete linkage clustergram based on\nweighted normalized unifrac distances", xlab="", ylab="distance", sub="")
```


## APPENDIX: Counts of high quality paired reads for each sample for the whole lane
```{r}
pander(read_counts)
```
